### PR TITLE
Makes Gitlab CI Interruptible

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR makes our Gitlab CI interruptible, similar to the setup that we have for github. Only CI that does not run on master/develop branch can be interrupted.

### CODE REVIEW

**Review by upsj — APPROVED** [review]
LGTM! Awesome that his now works :)

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -19,6 +19,13 @@ include:
   - local: '.gitlab/rules.yml'
   - local: '.gitlab/scripts.yml'
   - local: '.gitlab/variables.yml'
+  # This is a workaround to conditionally make the branch pipelines
+  # interruptible, because the flag does not directly support rules [1].
+  #
+  # [1] https://gitlab.com/gitlab-org/gitlab/-/issues/194023#note_1225906002
+  - local: '.gitlab/add-interrupt.yml'
+    rules:
+      - if: $CI_COMMIT_BRANCH != "master" && $CI_COMMIT_BRANCH != "develop"
```
Comment (yhmtsai) [inline .gitlab-ci.yml:28]: also the tagged branch `$CI_COMMIT_TAG` can not be interruptible.
We also use the tag for the release although the tagged version should only be pushed once.

**Comment by yhmtsai** [comment]
also @tcojean needs to adjust the option https://docs.gitlab.com/ee/ci/pipelines/settings.html#auto-cancel-redundant-pipelines to make it work

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -859,10 +867,10 @@ threadsanitizer:
...
     - LD_PRELOAD=/usr/local/lib/libomp.so
       CC=clang CXX=clang++
-        ctest -V -S cmake/CTestScript.cmake -DCTEST_BUILD_CONFIGURATION=TSAN
-        -DCTEST_MEMORYCHECK_TYPE=ThreadSanitizer
-        -DCTEST_MEMORYCHECK_SANITIZER_OPTIONS=ignore_noninstrumented_modules=1
-        --timeout 6000
+      ctest -V -S cmake/CTestScript.cmake -DCTEST_BUILD_CONFIGURATION=TSAN
+      -DCTEST_MEMORYCHECK_TYPE=ThreadSanitizer
+      -DCTEST_MEMORYCHECK_SANITIZER_OPTIONS=ignore_noninstrumented_modules=1
+      --timeout 6000
```
Comment (tcojean) [inline .gitlab-ci.yml:873]: please keep the alignment for all the scripts

**Comment by MarcelKoch** [comment]
I think the interrupt doesn't work if we force push... That makes it not as helpful with our workflow

**Comment by yhmtsai** [comment]
It's sad they do not handle the forced push. if it does not lead another error, it's still helpful partially.
I guess they go through all previous sha not the branch name.
Another annoying way is to use the API to get all pipeline with the same branch name and then cancel them by API.

**Comment by MarcelKoch** [comment]
rebase!

**Pull request merged**